OADP- 3949 -SSE-C Encryption backups by default (#75825)
openshift/openshift-docs: modules/oadp-ssec-encrypted-backups.adoc

// Module included in the following assemblies:
//
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-aws.adoc
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-azure.adoc
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-gcp.adoc
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-mcg.adoc
// * backup_and_restore/application_backup_and_restore/installing/installing-oadp-ocs.adoc

:_mod-docs-content-type: PROCEDURE
[id="oadp-ssec-encrypted-backups_{context}"]
= OADP SSE-C encrypted backups

> Amazon Web Services (AWS) S3 now applies server-side encryption with Amazon S3 managed keys (SSE-S3) as the base level of encryption for every bucket in Amazon S3. It is important to note that OpenShift API for Data Protection (OADP) encrypts data by using SSL/TLS, HTTPS, and the velero-repo-credentials secret when transferring the data off cluster to storage.

Reviewer: Please do not forget to use attributes

```suggestion
{aws-first} S3 applies server-side encryption with Amazon S3 managed keys (SSE-S3) as the base level of encryption for every bucket in Amazon S3. It is important to note that {oadp-first} encrypts data by using SSL/TLS, HTTPS, and uses the `velero-repo-credentials` secret when transferring the data from the cluster to storage.
```

> It is important to include an additional layer of encryption to prevent backed up data from becoming unencrypted in case of lost or stolen AWS credentials.

Reviewer: Using backed up data means it is a compound adjective, and should be hyphenated
```suggestion
It is important to include an additional layer of encryption to prevent backed-up data from becoming unencrypted in the event of lost or stolen AWS credentials.
```

> The _OpenShift velero-plugin-for-aws_ provides several additional encryption methods, review the configuration options and consider implementing additional encryption.

Reviewer: ```suggestion
The _OpenShift velero-plugin-for-aws_ provides several additional encryption methods; review the configuration options and consider implementing additional encryption.
```

> Server-side encryption with customer-provided keys (SSE-C), provides Red Hat users the ability to store your own encryption keys. This provides additional required security if the AWS credentials become exposed.

Reviewer: switch between second and third person

```suggestion
Server-side encryption with customer-provided keys (SSE-C), provides you with the ability to store your own encryption keys. This provides additional required security if the AWS credentials become exposed.
```

[WARNING]
====
It is important to secure cryptographic keys in a secure and safe manner. Encrypted data and backups will not be recoverable in the absence of the encryption key.
====

.Prerequisites / Known issues

> You might receive `credentials not found` errors. To avoid this issue you can create a workaround by adding a `volumeSnapshot` configuration to your `DPA` without specifying a credential as in the following example:

Reviewer: ```suggestion
You might receive `credentials not found` errors. To avoid this issue you can create a workaround by adding a `volumeSnapshot` configuration to your `DPA` without specifying a credential as in the following example:
```

        region: <region>
      provider: aws
# ...
----

.Procedure

You can configure an SSE-C encryption to offer additional security by applying the following steps.

. Create an encryption key by running the following command:
+
[source,terminal]
----
$ dd if=/dev/urandom bs=1 count=32 > sse.key
cat sse.key | base64 > sse_encoded.key
ln -s sse_encoded.key customer-key
----

. Create an OpenShift secret.

.. If you are initially installing and configuring OADP, you can create the aws credential and encryption key secret at the same time by running the following command:
+
[source,terminal]
----
$ oc create secret generic cloud-credentials --namespace openshift-adp --from-file cloud=<path>/openshift_aws_credentials,customer-key=<path>/sse_encoded.key
----

.. If you are updating an existing installation, edit the values of the `cloud-credential` `secret` block of the `DataProtectionApplication` CR manifest, as in the following example:
+
[source,yaml]
----
apiVersion: v1
data:
  cloud: W2Rfa2V5X2lkPSJBS0lBVkJRWUIyRkQ0TlFHRFFPQiIKYXdzX3NlY3JldF9hY2Nlc3Nfa2V5P<snip>rUE1mNWVSbTN5K2FpeWhUTUQyQk1WZHBOIgo=
  customer-key: v+<snip>TFIiq6aaXPbj8dhos=
kind: Secret
# ...
----

> . Edit the value of the `customerKeyEncryptionFile` attribute in the `backupLocations` block of `DataProtectionApplication` CR manifest, as in the following example:

Reviewer: ```suggestion
. Edit the value of the `customerKeyEncryptionFile` attribute in the `backupLocations` block of the `DataProtectionApplication` CR manifest, as in the following example:
```

+
[source,yaml]
----
spec:
  backupLocations:
    - velero:
        config:
          customerKeyEncryptionFile: /credentials/customer-key
          profile: default
# ...
----

[WARNING]
====
Existing installations require you to restart the Velero pod to remount the secret credentials properly.
====

> The installation is complete, and you can back up and restore OpenShift resources. The data saved in AWS s3 storage is encrypted with the new key and cannot be downloaded from the AWS S3 console or API without the additional encryption key.

Reviewer: ```suggestion
The installation is complete, and you can back up and restore OpenShift resources. The data saved in AWS s3 storage is encrypted with the new key and you cannot download it from the AWS S3 console or API without the additional encryption key
```

.Verification

> Test to verify that encrypted files cannot be downloaded without the additional key.

Reviewer: ```suggestion
The purpose of this test is to verify that encrypted files cannot be downloaded without the inclusion of an additional key.
```

. Create a test file by running the following command:

+
[source,terminal]
----
$ echo "encrypt me please" > test.txt
----

. Upload the test file by running the following command:

+
[source,terminal]
----
$ aws s3api put-object \
  --bucket <bucket> \
  --key test.txt \
  --body test.txt \
  --sse-customer-key fileb://sse.key \
  --sse-customer-algorithm AES256
----

> . Try to download the file, which should fail because the file is now encrypted with an additional key, by using the S3 web console or the terminal by running the following command:

Reviewer: ```suggestion
. Try to download the file, this should fail because the file is now encrypted with an additional key. By using the S3 web console or the terminal, issue the following command:
```

+
[source,terminal]
----
$ s3cmd get s3://<bucket>/test.txt test.txt
----

> . Download the file with the additional encryption key by running the following commands:

Reviewer: ```suggestion
. Download the file with the additional encryption key by issuing the following commands:
```

Author: https://docs.redhat.com/en/documentation/openshift_container_platform/4.15/html-single/backup_and_restore/index#oadp-using-ca-certificates-with-velero-command-aliased-for-velero-deployment_installing-oadp-aws

OCP prefers we write "by running the following command."

+
[source,terminal]
----
$ aws s3api get-object \
    --bucket <bucket> \
    --key test.txt \
    --sse-customer-key fileb://sse.key \
    --sse-customer-algorithm AES256 \
    downloaded.txt
----
+
[source,terminal]
----
$ cat downloaded.txt
encrypt me please
----

> . You can run the same download call on velero backed up files by running the following command:

Reviewer: ```suggestion
. You can run the same download call on the files backed up with Velero by issuing the following command:
```
